ports/esp32: Add support for SPI-based ethernet chips

damz opened the pull request:
### What?

Add support for various SPI-based ethernet chips (W5500, KSZ8851SNL, DM9051) to the ESP32 port.

The impetus for this was to support the [POE Featherwing](https://www.crowdsupply.com/silicognition/poe-featherwing), which is based on the Wiznet W5500 chip.

### How?

Leverage the existing support in ESP-IDF for these chips, and the existing support for `network.LAN` in the ESP32 port.

In particular, this doesn't leverage the wiznet5k support that is used on the rp2 and stm32 ports.

#### A note on testing

I tested this on the [POE Featherwing](https://www.crowdsupply.com/silicognition/poe-featherwing) (with the SJIRQ solder jumper bridged) and a [ESP32-S3 feather](https://www.adafruit.com/product/5477).

While I do not have access to hardware using the other chips, the set up method for them comes from the [ethernet/basic](https://github.com/espressif/esp-idf/blob/release/v4.4/examples/ethernet/basic/main/ethernet_example_main.c) example from ESP-IDF 4.4, so I am relatively confident that it is correct.

#### A note about the interrupt pin

The W5500 implementation within ESP-IDF relies on hardware interrupt, and requires the interrupt pin from the W5500 to be wired to a GPIO. This is not the case by default on the [Adafruit Ethernet FeatherWing](https://www.adafruit.com/product/3201), which makes it not directly compatible with this implementation.

beyonlo commented:
@damz Hello.

I don't know if this can help you, but maybe [this](https://github.com/micropython/micropython/pull/8540#issuecomment-1151701493) has a very good relation with this PR. The `Wiznet` ethernet `SPI` chips already works on `RP2040` and `STM,` so, maybe to port that to `ESP32` is a good idea to be a common base for many ports!

beyonlo commented:
@damz 

> In particular, this doesn't leverage the wiznet5k support that is used on the rp2 and stm32 ports.

I'm sorry, you already made it clear that this PR implementation does not use that implementation that I linked before.

Thank you.

stephanelsmith commented:
Love this PR.  Would be a great addition, especially with ESP-IDF 4.4 underlying support.

dpgeorge commented on ports/esp32/modnetwork.c:
```diff
@@ -229,7 +229,7 @@ STATIC const mp_rom_map_elem_t mp_module_network_globals_table[] = {
     #endif
 
-    #if (ESP_IDF_VERSION_MAJOR == 4) && (ESP_IDF_VERSION_MINOR >= 1) && (CONFIG_IDF_TARGET_ESP32)
+    #if MODNETWORK_LAN
```
This config macro should match others and be `MICROPY_PY_NETWORK_LAN`.

damz replied:
Done.

dpgeorge commented on ports/esp32/mpconfigport.h:
```diff
@@ -217,3 +217,21 @@ typedef long mp_off_t;
+#if MODNETWORK_LAN && CONFIG_ETH_USE_SPI_ETHERNET
+#ifndef NETWORKMOD_SPI_CLOCK_SPEED_MZ
+#if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32C2
+#define NETWORKMOD_SPI_CLOCK_SPEED_MZ       (12)
```
This should be something like `MICROPY_PY_NETWORK_LAN_SPI_CLOCK_SPEED_MZ`.

damz replied:
Done.

dpgeorge commented on ports/esp32/network_lan.c:
```diff
@@ -97,14 +102,17 @@ STATIC mp_obj_t get_lan(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_ar
         { MP_QSTR_power,        MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
         { MP_QSTR_phy_addr,     MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_INT },
         { MP_QSTR_phy_type,     MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_INT },
+        { MP_QSTR_spi_id,       MP_ARG_INT, {.u_int = -1} },
```
I think this should accept a `machine.SPI` object (not an int) to be consistent with args like `mdc` which accept a Pin object.

damz replied:
Makes sense. I renamed the argument to just `spi` while I was at it.

damz committed:
esp32/network_lan: Fix setting MAC address of ethernet devices.

`esp_eth_ioctl(ETH_CMD_S_MAC_ADDR)` sets the MAC address of the hardware
device, but we also need to notify the upper layers of the change so that
e.g. DHCP work properly.

damz committed:
esp32/network_lan: Add support for SPI-based ethernet chips.

Add support for various SPI-based ethernet chips (W5500, KSZ8851SNL,
DM9051) to the ESP32 port.  This leverages the existing support in ESP-IDF
for these chips -- which configures these chips in "MAC raw" mode -- and
the existing support for network.LAN in the ESP32 port.  In particular,
this doesn't leverage the wiznet5k support that is used on the rp2 and
stm32 ports (because that's for native use of lwIP).

Tested on the POE Featherwing (with the SJIRQ solder jumper bridged) and a
ESP32-S3 feather.

A note about the interrupt pin: The W5500 implementation within ESP-IDF
relies on hardware interrupt, and requires the interrupt pin from the W5500
to be wired to a GPIO.  This is not the case by default on the Adafruit
Ethernet FeatherWing, which makes it not directly compatible with this
implementation.

damz commented:
@dpgeorge Thanks for the review. I updated the PR.

dpgeorge commented on ports/esp32/network_lan.c:
```diff
@@ -97,14 +102,17 @@ STATIC mp_obj_t get_lan(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_ar
-        { MP_QSTR_mdc,          MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_OBJ },
-        { MP_QSTR_mdio,         MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_mdc,          MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+        { MP_QSTR_mdio,         MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
```
Now that these are not required args, what happens when you don't specify them but they are needed (in the original case of using a non-SPI PHY)?  Does it give a sensible error message?

damz replied:
Good point, it wouldn't. I added a manual check for this in a !IS_SPI_PHY path.

damz replied:
I adjusted the code flow a bit by moving the ESP32 embedded ethernet MAC initialization at the bottom, I think it makes more sense now. (And from the look of it, it will make porting to ESP-IDF 5 simpler.)

robert-hh replied:
What is the relation to PRs  #10220 and #10329 ? I see that it is partially overlapping. The referenced PR's allow in addition to set the ref_clk_pin and ref_clk_mode.

damz replied:
@robert-hh It's just moving code around. It doesn't conceptually clash with #10329 at all (but it will likely require a small manual conflict fix).

damz commented:
@dpgeorge I think all your comments so far have been addressed.

dpgeorge commented:
I rebased this on latest master.

dpgeorge commented:
Thanks @damz for this, it's a good enhancement!

beyonlo commented:
Hello

I'm very happy that we are closer to have SPI-based ethernet chips really supported on the ESP32 family! That feature is something amazing for many projects!

I have some questions:

>Add support for various SPI-based ethernet chips (W5500, KSZ8851SNL, DM9051) to the ESP32 port.
1. About the `Wiznet` chip, will works just with the `W5500`, or also others `5K` family chips will be supported, like as the `W5100S`?
2. Will that ethernet chips works as mac raw mode? I mean, the limits of sockets for example, `SSL`, etc, will works by ESP-IDF, not from chip limits?
3. Everythink that works on IP using the WiFi, will works via Ethernet? I mean, SSL, and so on. I understand that as that support are in ESP-IDF, should be a transparent way and everything works as on the WiFi.
4. Will be possible to use at the same time WiFi and Ethernet? I mean, I will have two interfaces (`IPs`) running together.

Thank you in advance!

licsber commented:
Hello

I test `esp32-20230426-v1.20.0.bin` with `ESP32-WROOM-32E` chip, but when i import network, there is no `PHY_W5500` enum. Just get `DP83848 IP101 LAN8710 LAN8720 RTL8201`.

Is there any examples for using SPI-based Ethernet chips ? Does i need compile firmware myself?

Also test nightly firmware `esp32-20230610-unstable-v1.20.0-189-gfd277704c.bin`, no help.
